File: Sources/Views/AzureToastView.m
#import "AzureToastView.h"


@implementation AzureToastView {

	UIView *toastView;
	NSLayoutConstraint *bottomAnchorConstraint;

}

- (id)init {

	self = [super init];

	if(!self) return nil;

	[self setupToastView];

	[NSNotificationCenter.defaultCenter removeObserver:self];
	[NSNotificationCenter.defaultCenter addObserver:self selector:@selector(fadeInOutCopyPinToastView) name:@"fadeInOutToast" object:nil];

	return self;

}


- (void)setupToastView {

	self.translatesAutoresizingMaskIntoConstraints = NO;

	toastView = [UIView new];
	toastView.alpha = 0;
	toastView.backgroundColor = kAzureMintTintColor;
	toastView.layer.cornerCurve = kCACornerCurveContinuous;
	toastView.layer.cornerRadius = 20;
	toastView.translatesAutoresizingMaskIntoConstraints = NO;
	[self addSubview: toastView];

	toastViewLabel = [UILabel new];
	toastViewLabel.font = [UIFont systemFontOfSize: 14];
	toastViewLabel.textColor = UIColor.labelColor;
	toastViewLabel.numberOfLines = 0;
	toastViewLabel.textAlignment = NSTextAlignmentCenter;
	toastViewLabel.adjustsFontSizeToFitWidth = YES;
	toastViewLabel.translatesAutoresizingMaskIntoConstraints = NO;
	[toastView addSubview: toastViewLabel];

	[toastView.centerXAnchor constraintEqualToAnchor: self.centerXAnchor].active = YES;
	bottomAnchorConstraint = [toastView.bottomAnchor constraintEqualToAnchor: self.safeAreaLayoutGuide.bottomAnchor constant : 50];
	bottomAnchorConstraint.active = YES;
	[toastView.widthAnchor constraintEqualToConstant: 120].active = YES;
	[toastView.heightAnchor constraintEqualToConstant: 40].active = YES;

	[toastViewLabel.centerXAnchor constraintEqualToAnchor: toastView.centerXAnchor].active = YES;
	[toastViewLabel.centerYAnchor constraintEqualToAnchor: toastView.centerYAnchor].active = YES;
	[toastViewLabel.leadingAnchor constraintEqualToAnchor: toastView.leadingAnchor constant: 10].active = YES;
	[toastViewLabel.trailingAnchor constraintEqualToAnchor: toastView.trailingAnchor constant: -10].active = YES;

}


- (void)fadeInOutCopyPinToastView {

	toastViewLabel.text = @"Copied!";
	[self fadeInOutToastViewWithFinalDelay: 0.2];

}


- (void)fadeInOutToastViewWithFinalDelay:(CGFloat)delay {

	[UIView animateViewWithDelay:0 withAnimations:^ {

		[self animateToastViewWithConstraintConstant: -20 andAlpha: 1];

	} withCompletion:^(BOOL finished) {

		[UIView animateViewWithDelay:0.2 withAnimations:^ {

			[UIView makeRotationTransformForView:toastView andLabel:toastViewLabel];
			[self layoutIfNeeded];

		} withCompletion:^(BOOL finished) {

			[UIView animateViewWithDelay:delay withAnimations:^ {

				[self animateToastViewWithConstraintConstant: 50 andAlpha: 0];

			} withCompletion:^(BOOL finished) {

				toastView.layer.transform = CATransform3DIdentity;
				toastViewLabel.layer.transform = CATransform3DIdentity;

			}];

		}];

	}];

}


- (void)animateToastViewWithConstraintConstant:(CGFloat)constant andAlpha:(CGFloat)alpha {

	bottomAnchorConstraint.constant = constant;
	toastView.alpha = alpha;
	[self layoutIfNeeded];

}

@end

File: Sources/Views/UI/AzureToastView.swift
import UIKit


final class AzureToastView: UIView {

	private var bottomAnchorConstraint: NSLayoutConstraint?

	private lazy var toastView: UIView = {
		let view = UIView()
		view.alpha = 0
		view.backgroundColor = .kAzureMintTintColor
		view.layer.cornerCurve = .continuous
		view.layer.cornerRadius = 20
		addSubview(view)
		return view
	}()

	private lazy var toastViewLabel: UILabel = {
		let label = UILabel()
		label.font = .systemFont(ofSize: 14)
		label.textColor = .label
		label.numberOfLines = 0
		label.textAlignment = .center
		label.adjustsFontSizeToFitWidth = true
		toastView.addSubview(label)
		return label
	}()

	required init?(coder: NSCoder) {
		super.init(coder: coder)
	}

	override init(frame: CGRect) {
		super.init(frame: frame)
		setupToastView()
	}

	// ! Private

	private func setupToastView() {
		translatesAutoresizingMaskIntoConstraints = false

		setupHorizontalConstraints(forView: toastView, leadingConstant: 10, trailingConstant: -10)
		toastView.heightAnchor.constraint(equalToConstant: 40).isActive = true

		toastView.centerViewOnBothAxes(toastViewLabel)
		toastView.setupHorizontalConstraints(forView: toastViewLabel, leadingConstant: 10, trailingConstant: -10)

		let guide = safeAreaLayoutGuide
		bottomAnchorConstraint = toastView.bottomAnchor.constraint(equalTo: guide.bottomAnchor, constant: 50)
		bottomAnchorConstraint?.isActive = true
	}

	private func fadeInOutToastView(withFinalDelay delay: TimeInterval) {
		animateView(animations: {
			self.animateToastView(withConstraintConstant: -20, alpha: 1)
		}, completion: { _ in
			self.animateView(withDelay: 0.2, animations: {
				self.makeRotationTransform(forViews: [self.toastView, self.toastViewLabel])
			}, completion: { _ in
				self.animateView(withDelay: delay, animations: {
					self.animateToastView(withConstraintConstant: 50, alpha: 0)
				}, completion: { _ in
					self.toastView.layer.transform = CATransform3DIdentity
					self.toastViewLabel.layer.transform = CATransform3DIdentity
				})
			})
		})
	}

	private func animateToastView(withConstraintConstant constant: CGFloat, alpha: CGFloat) {
		bottomAnchorConstraint?.constant = constant
		toastView.alpha = alpha
		layoutIfNeeded()
	}

}

extension AzureToastView {

	// ! Public

	func fadeInOutToastView(withMessage message: String, finalDelay delay: TimeInterval) {
		toastViewLabel.text = message
		fadeInOutToastView(withFinalDelay: delay)
	}

}
